model/src/main/java/org/eclipse/rdf4j/model/util/Models.java

```diff
@@ -724,51 +717,77 @@ private static boolean statementsMatch(Statement st1, Statement st2, Map<Resourc
 			return false;
 		}
 
-		if (isBlank(context1) && isBlank(context2)) {
-			Resource mappedBNode = bNodeMapping.get(context1);
+		if (bnodeValueMatching(bNodeMapping, context1, context2)) {
+			return false;
+		}
+
+		return true;
+	}
+
+	private static boolean bnodeValueMatching(Map<Resource, Resource> bNodeMapping, Value obj1, Value obj2) {
+		if (isBlank(obj1) && isBlank(obj2)) {
+			Resource mappedBNode = bNodeMapping.get(obj1);
 
 			if (mappedBNode != null) {
-				// bNode 'context1' was already mapped to some other bNode
-				if (!context2.equals(mappedBNode)) {
-					// 'context1' and 'context2' do not match
-					return false;
-				}
+				// bNode 'obj1' was already mapped to some other bNode
+				// 'obj1' and 'obj2' do not match
+				return !obj2.equals(mappedBNode);
 			} else {
-				// 'context1' was not yet mapped. we need to check if 'context2' is
-				// a
+				// 'obj1' was not yet mapped. we need to check if 'obj2' is a
 				// possible mapping candidate
-				if (bNodeMapping.containsValue(context2)) {
-					// 'context2' is already mapped to some other value.
-					return false;
-				}
+				// 'obj2' is already mapped to some other value.
+				return bNodeMapping.containsValue(obj2);
 			}
 		} else {
-			// contexts are not (both) bNodes
-			if (!context1.equals(context1)) {
-				return false;
-			}
+			// objects are not (both) bNodes
+			return !obj1.equals(obj2);
 		}
-
-		return true;
 	}
 
 	private static boolean isBlank(Value value) {
+		if (value instanceof BNode) {
```

Comment:
Checking for instanceOfBnode first also gave a nice improvement. I think it might be mostly due to the jvm being able to do better escape analysis, but that's just a hunch from looking at the flight recordings.